stdlib/eq-paths.mc

```diff
@@ -0,0 +1,363 @@
+include "digraph.mc"
+include "string.mc"
+
+-- This file implements eqPaths: computing equivalence paths for decision
+-- points.
+
+-- Input: a graph, a start node v and a depth d
+-- Output: the set of equivalence paths ending in v. The lengths of the
+-- paths are at most d.
+
+-- It's really a graph algorithm, but very specific, and the test cases
+-- takes lots of space, so it's currently in a separate file from digraph.mc.
+
+let eqPaths = lam g. lam v. lam d.
```

Comment:
You are right that it's inconsistent in a sense, as the following graph g1 with direct recursion:
digraph {
    1->2[label=l1];
    2->2[label=l2];
    2->4[label=l4];
}
would give eqpaths g1 4 4 = [l1, l4] (source 4, depth 4)
while this similar graph g2 with indirect recusion 
digraph {
    1->2[label=l1];
    2->3[label=l2];
    3->2[label=l3]
    2->4[label=l4];
}
would give eqpaths g2 4 4 = [l3, l4]. 
To make it consistent perhaps the result should be [l1,l4] in the latter case as well.

The source of the inconsistency above was to be able to distinguish between "in a mutual recursion" and "not in a mutual recursion" for nodes in a mutual recursion. That is, if node 2 above is the source, we wanted:
eqpaths g2 2 1 = [l3]
eqpaths g2 2 2 = [l3] 
But now I don't feel sure if this makes sense. Perhaps always cutting the cycle if it is completed is the best as it is the most consistent way. In that way, we would have:
eqpaths g2 2 1 = [l3]
eqpaths g2 2 2 = [l1]

Reply:
![g1](https://user-images.githubusercontent.com/14043711/85837517-65db3c00-b798-11ea-87fa-0f1db7eaf36c.png)
![g2](https://user-images.githubusercontent.com/14043711/85837526-68d62c80-b798-11ea-8a98-2dcdf6febd36.png)

Reply:
Sorry, I got confused and missed some paths in the examples above. It has been updated now.

Reply:
[I'm guessing you could put the ascii graphs in a code block to render appropriately, but I'm not sure]

I really have no idea which way it should be. Maybe it's ok that the number of equivalence classes can decrease (which I would otherwise consider as surprising) as the depth gets greater, as we discover new mutual recursions? Either way that's probably a question for later.